File: solutions/elixir/hamming/1/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2) do
    {:error, "strands must be of equal length"}
  end

  def hamming_distance(strand1, strand2) do
    distance = 
      Enum.zip(strand1, strand2)
      |> Enum.reduce(0, fn ({l1, l2}, acc) -> if(l1 != l2, do: acc+1, else: acc) end)
    {:ok, distance}
  end
end

File: solutions/go/hamming/1/hamming.go
package hamming

import "errors"

func Distance(a, b string) (int, error) {
    aLetters := []rune(a)
    bLetters := []rune(b)
	if len(aLetters) != len(bLetters) {
        return 0, errors.New("DNA strands must have same length")
    }
    dist := 0
    for i := 0; i < len(aLetters); i++ {
        if aLetters[i] != bLetters[i] {
            dist++
        }
    }
    return dist, nil
}
